src/graph.rs

```diff
@@ -86,6 +92,131 @@ impl<'tree> Graph<'tree> {
+                if !node.outgoing_edges.is_empty() {
+                    let edges = Vec::from_iter(
+                        node.outgoing_edges
+                            .iter()
+                            .map(|(sink, edge)| JSONEdge(sink, edge, ctx)),
+                    );
+                    map.serialize_entry("edges", &edges)?;
+                }
+                if !node.attributes.values.is_empty() {
+                    map.serialize_entry("attrs", &JSONAttributes(&node.attributes, ctx))?;
+                }
+                map.end()
+            }
+        }
+
+        impl<'a> ser::Serialize for JSONEdge<'a> {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+            where
+                S: serde::Serializer,
+            {
+                let sink = self.0;
+                let edge = self.1;
+                let ctx = self.2;
+                let mut map = serializer.serialize_map(None)?;
+                map.serialize_entry("sink", &sink)?;
+                if !edge.attributes.values.is_empty() {
+                    map.serialize_entry("attrs", &JSONAttributes(&edge.attributes, &ctx))?;
+                }
+                map.end()
+            }
+        }
+
+        impl<'a> ser::Serialize for JSONAttributes<'a> {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+            where
+                S: serde::Serializer,
+            {
+                let attrs = self.0;
+                let ctx = self.1;
+                let mut map = serializer.serialize_map(None)?;
+                for (_, (key, value)) in attrs.values.iter().enumerate() {
+                    map.serialize_entry(&JSONIdentifier(key, ctx), &JSONValue(value, ctx))?;
+                }
+                map.end()
+            }
+        }
+
+        impl<'a> ser::Serialize for JSONIdentifier<'a> {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+            where
+                S: serde::Serializer,
+            {
+                let identifier = self.0;
+                let ctx = self.1;
+                let symbol = ctx.identifiers.resolve(identifier.0).unwrap();
+                serializer.serialize_str(symbol)
+            }
+        }
+
+        impl<'a> ser::Serialize for JSONValue<'a> {
+            fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+            where
+                S: serde::Serializer,
+            {
+                let value = self.0;
+                let ctx = self.1;
+                match value {
+                    Value::Null => serializer.serialize_none(),
+                    Value::Boolean(bool) => serializer.serialize_bool(*bool),
+                    Value::Integer(integer) => serializer.serialize_u32(*integer),
+                    Value::String(string) => serializer.serialize_str(string),
+                    // FIXME: there's no way to distinguish sets and lists, so we can't roundtrip accurately
```

Comment:
I think I probably want to use one of the variant serialization options to handle this; it will also catch the various integer cases (below).

Reply:
I assume that would produce some form of tagged output? I think that makes sense for values.

Reply:
Yeah, that's what I'm thinking. I'd like to merge this without those for the moment if folks are cool with that (what's the release process like on this repo?); it's not that this change will take a long time, more that it will _definitely_ interfere with work on the `semantic-analysis` side, and I'd like to postpone that interference for a week or so.

Reply:
> what's the release process like on this repo?

ad hoc, need to write a `script` for it.  Nothing automatic, so it's fine to have this merged and unreleased for a period of time.